### PR TITLE
ipc: notifications refactor

### PR DESCRIPTION
**Pull request by tlauda**
This set of patches changes the way notifications are allocated and sent. More details in commit descriptions.

### CODE REVIEW

File: src/drivers/imx/ipc.c [hunk 1]
```diff
@@ -23,6 +23,7 @@
 #include <ipc/topology.h>
```
Comment (dbaluta) [inline src/drivers/imx/ipc.c]: @tlauda do we have a real usecase where checking IPC bits is not enough?

**Comment by tlauda** [comment]
> Looks good, but my only thought is that an allocation failure may prevent an IPC message ? Should we pre-allocate the IPC messages (i.e embed into component private data). That way we are always guaranteed to succeed, I would also pre-alloc 8 messages on IPC core too so that we always **copy** client messages into IPC core messages (meaning we wont have collisions between host/IPC core and clients).

IPCs are allocated along with notification owner. There are no other allocations during the owner's lifetime. I think I don't get the second part. What is your definition of host/IPC core/client?

File: src/drivers/imx/ipc.c (hunk 1 above)
Comment (tlauda) [reply src/drivers/imx/ipc.c]: @dbaluta Yes, during specific power flows on cAVS DSP.

**Comment by mmaka1** [comment]
@lgirdwood Pre-allocation on the IPC core would consume a lot of memory when max ipc size is aligned with the mailbox size since the core would need to alloc max for each one in the pool. While an originator can do precise allocation.

File: src/ipc/handler.c [hunk 2]
```diff
@@ -1371,158 +1306,94 @@ void ipc_cmd(struct sof_ipc_cmd_hdr *hdr)
...
-	default:
-		/* not found */
-		return NULL;
-	}
+	ipc_msg_queue(msg, event->rhdr.hdr.cmd, event, true);
 }
 
-void ipc_prepare_host_message(struct ipc_msg *msg, uint32_t header,
-			      void *tx_data, size_t tx_bytes)
+void ipc_xrun_msg_send(struct ipc_msg *msg, const struct comp_dev *dev,
```
Comment (mmaka1) [inline src/ipc/handler.c]: I am not sure why there is a dedicated function needed for each specific message. Looks like some generic ipc_prepare(...) called by a client could do the work?

File: src/ipc/handler.c [hunk 3]
```diff
@@ -1371,158 +1306,94 @@ void ipc_cmd(struct sof_ipc_cmd_hdr *hdr)
...
 {
-	int ret;
+	posn->rhdr.hdr.cmd = SOF_IPC_GLB_STREAM_MSG | SOF_IPC_STREAM_TRIG_XRUN |
+		dev_comp_id(dev);
+	posn->rhdr.hdr.size = sizeof(*posn);
+	posn->comp_id = dev_comp_id(dev);
 
-	msg->header = header;
-	msg->tx_size = tx_bytes;
+	mailbox_stream_write(dev->pipeline->posn_offset, posn, sizeof(*posn));
```
Comment (mmaka1) [inline src/ipc/handler.c]: This should not be part of ipc code I believe. How about:
```c
position_update(...)
{
    mailbox_stream_write(...)
    ipc_send(...)
}
```

**Comment by lgirdwood** [comment]
@tlauda the thinking here with hostIPC core vs client is client changing any data during a copy to SRAM window (e.g. during client IRQ code).
@mmaka1 Sorry to be clear, I was thinking more about the struct msg pre-allocation, rather than the whole mailbox.

**Comment by tlauda** [comment]
@lgirdwood With these set of patches we're totally changing the approach to owning and sending notifications. The idea was proposed by @mmaka1 as a potential solution for optimizing memory usage. Just to sum up:
1. There is no longer any common pool for IPC notifications. Every initiator allocates one notification message with exact size it requires.
2. Notifications can be send with high priority meaning we will try to send it right away and if the upstream channel is busy, we will add it to the beginning of message queue.
3. If it happens the notification from this particular originator is already in the queue, we just update the data and don't insert it again.
4. Everything is protected using lock.

**Comment by paulstelian97** [comment]
> @lgirdwood With these set of patches we're totally changing the approach to owning and sending notifications. The idea was proposed by @mmaka1 as a potential solution for optimizing memory usage. Just to sum up:
> 
> 1. There is no longer any common pool for IPC notifications. Every initiator allocates one notification message with exact size it requires.
> 2. Notifications can be send with high priority meaning we will try to send it right away and if the upstream channel is busy, we will add it to the beginning of message queue.
> 3. If it happens the notification from this particular originator is already in the queue, we just update the data and don't insert it again.
> 4. Everything is protected using lock.

I have one issue: should there be support for canceling a notification that is in the queue whenever the data structures for a component are freed? We 100% don't want to send invalid notifications via a use-after-free do we?

**Comment by tlauda** [comment]
> > @lgirdwood With these set of patches we're totally changing the approach to owning and sending notifications. The idea was proposed by @mmaka1 as a potential solution for optimizing memory usage. Just to sum up:
> > 
> > 1. There is no longer any common pool for IPC notifications. Every initiator allocates one notification message with exact size it requires.
> > 2. Notifications can be send with high priority meaning we will try to send it right away and if the upstream channel is busy, we will add it to the beginning of message queue.
> > 3. If it happens the notification from this particular originator is already in the queue, we just update the data and don't insert it again.
> > 4. Everything is protected using lock.
> 
> I have one issue: should there be support for canceling a notification that is in the queue whenever the data structures for a component are freed? We 100% don't want to send invalid notifications via a use-after-free do we?

There is such support in ipc_msg_free().

**Comment by lgirdwood** [comment]
Jenkins CI is known issue, but we have UT failure in the internal CI.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
Rerunning CI now that WHL DUT has been rebooted.

**Pull request merged**